TechnicalDocuments/Proposals/BuildersToFunctions.md

```diff
@@ -0,0 +1,161 @@
+    enum Alert {
+            static func makeAlert(for error: CoreError, dismiss: (() -> Void)?) -> UIAlertController
+            static func makeAlert(for message: String, title: String?, dismiss: (() -> Void)?) -> UIAlertController
+            static func makeAlert(for error: AlertError,
+                                  primaryAction: UIAlertAction,
+                                  secondaryAction: UIAlertAction) -> UIAlertController
+    }
+}
+```
+
+This would allow us to remove the protocol `AlertBuilder` and remove all the compliances across the codebase. From a usage point of view:
+
+```swift
+final class AppointmentListFlowController {
+    let primary: Flow
+    let secondary: Flow
+
+    init(primary: Flow, secondary: Flow) {
+        self.primary = primary
+        self.secondary = secondary
+    }
+
+    func handle(_ route: AppointmentListViewModel.Route) {
+        switch route {
+        case let .showAlert(error):
+            Builder.Alert.makeAlert(for: error, dismiss: secondary.dismiss)
+                |> secondary.present
+        }
+    }
+}
+```
+
+I will provide further examples in the proposal
+
+## Motivation
+
+There are a couple of reasons for this change:
+
+1. It was decided that DI via `Current` would happen at the builder level. This means that the requirement to inject dependencies via the initializer, in builders, ceases to exist. This would: remove protocols, clean FlowController's APIs and dissolve the concept of child builders. 
+2. It would become easier to find factory methods to construct screens. One could just type `Builder.` and Xcode would autocomplete with the available factory methods. It would also free engineers from wasting too much time finding dependencies if they are immediately accessible via `Current` and the namespaced free functions. This is not a justification for engineers to not understand the codebase and how to do things "properly". It is simply an observation of how cumbersome passing dependencies via initializer is, the raison d'être for `Current` and the overall way of creating screens (VM + VC + FC). 
+
+The first point is about simplicity and clean-up of multiple ways of creating screens, while the second point focuses on discoverability and development ergonomics.   
+
+## Proposed solution
+
+**Note:** The proposed solution assumes that a way to authenticate HTTP requests is available at `Current` level (in this case something in the shape of `MainUserSession` abstracted via `UserSession`).
+
+Let's use `MedicalHistoryBuilder.swift` as an example, since it offers a trivial API: 
+
+``` swift
+init(session: UserSession, visuals: VisualDependenciesProtocol)
+```
+
+And the corresponding public API:
+
+```swift
+func make(modal: Flow) -> UIViewController
+```
+
+Given the Module where it's located it could be migrated to:
+
+```swift
+extension Builder {
+    enum ClinicalRecords {
+        static func makeMedicalHistory(modal: Flow) -> UIViewController
```

Comment:
I wander if it make to have them as `var` closures? So you would be able stub them if we start to unit test `FlowControllers`.

Reply:
How would you make the default value to use `Current`?

Reply:
```
enum Builder {
    static var buildVC: () -> UIViewController = {
        let analitics = Current.analyticsService
        return UIViewController()
    }
}
```

Reply:
But the issue now is that it's mutable 🙈

Reply:
@sergdort It's fine and exactly the same how we agreed to use `Current` in builders. It's mutable now but we can just change it to `public internal(set) var Current: World = .production` to allow mutation only from tests using `@tesable import`